iceoryx_hoofs/source/posix_wrapper/named_semaphore.cpp

```diff
@@ -0,0 +1,221 @@
+{
+    auto result = posixCall(iox_sem_open)(createNameWithSlash(name).c_str(), 0)
+                      .failureReturnValue(IOX_SEM_FAILED)
+                      .ignoreErrnos(ENOENT)
+                      .evaluate();
+
+    if (result.has_error())
+    {
+        switch (result.get_error().errnum)
+        {
+        case EACCES:
+            LogError() << "Insufficient permissions to open semaphore \"" << name << "\".";
+            return cxx::error<SemaphoreError>(SemaphoreError::PERMISSION_DENIED);
+        case EMFILE:
+            LogError() << "The per-process limit of file descriptor exceeded while opening the semaphore \"" << name
+                       << "\"";
+            return cxx::error<SemaphoreError>(SemaphoreError::FILE_DESCRIPTOR_LIMIT_REACHED);
+        case ENFILE:
+            LogError() << "The system wide limit of file descriptor exceeded while opening the semaphore \"" << name
+                       << "\"";
+            return cxx::error<SemaphoreError>(SemaphoreError::FILE_DESCRIPTOR_LIMIT_REACHED);
+        case ENOMEM:
+            LogError() << "Insufficient memory to open the semaphore \"" << name << "\".";
+            return cxx::error<SemaphoreError>(SemaphoreError::OUT_OF_MEMORY);
+        default:
+            LogError() << "This should never happen. An unknown error occurred while opening the semaphore \"" << name
+                       << "\".";
+            return cxx::error<SemaphoreError>(SemaphoreError::UNDEFINED);
+        }
+    }
+
+    if (result->errnum != ENOENT)
+    {
+        constexpr bool HAS_OWNERSHIP = false;
+        uninitializedSemaphore.emplace(result.value().value, name, HAS_OWNERSHIP);
+        return cxx::success<bool>(true);
+    }
+
+    return cxx::success<bool>(false);
+}
+
+static cxx::expected<SemaphoreError> createSemaphore(cxx::optional<NamedSemaphore>& uninitializedSemaphore,
+                                                     const NamedSemaphore::Name_t& name,
+                                                     const OpenMode openMode,
+                                                     const cxx::perms permissions,
+                                                     const uint32_t initialValue) noexcept
+{
+    auto result = posixCall(iox_sem_open_ext)(createNameWithSlash(name).c_str(),
+                                              convertToOflags(openMode),
+                                              static_cast<mode_t>(permissions),
+                                              static_cast<unsigned int>(initialValue))
+                      .failureReturnValue(IOX_SEM_FAILED)
+                      .evaluate();
+
+    if (result.has_error())
+    {
+        switch (result.get_error().errnum)
+        {
+        case EACCES:
+            LogError() << "Insufficient permissions to create semaphore \"" << name << "\".";
+            return cxx::error<SemaphoreError>(SemaphoreError::PERMISSION_DENIED);
```

Comment:
Does this really need to be duplicated? There are already other posix wrapper which handle all of this in a single method.

Reply:
Actually it is not duplicated. The error message state the internal context if one would like to create, remove or open a semaphore.
Furthermore not every posix call has the same errno's. So when I put them all in one function then I handle errnos for calls which never would set them.
But what happens when on some platform now this errno's is set, with a generic error handling this could lead to confusion.

Example:
`sem_post` would set the errno `EPERM` then the generic error handling would state no permission to do whatever you wanna do. But this is not allowed in the posix standard and most likely an error in the platform.


But I see the risks here as quite minimal. So I would be fine to move this all into one separate big error handling function - @elBoberido it is your decision.

Reply:
Well, it's up to you. It just looked like an ideal candidate for code reuse.

To stay with your example, how do I know that I got an `EPERM` when the error message only tells me about an `unknown error`. Could you please also print that value?

Reply:
I take everything back I said